botcore/_bot.py

```diff
@@ -0,0 +1,265 @@
+        """Clear the internal guild available event when self.guild_id becomes unavailable."""
+        if guild.id != self.guild_id:
+            return
+
+        self._guild_available.clear()
+
+    async def on_guild_available(self, guild: discord.Guild) -> None:
+        """
+        Set the internal guild available event when self.guild_id becomes available.
+
+        If the cache appears to still be empty (no members, no channels, or no roles), the event
+        will not be set and `guild_available_but_cache_empty` event will be emitted.
+        """
+        if guild.id != self.guild_id:
+            return
+
+        if not guild.roles or not guild.members or not guild.channels:
+            msg = "Guild available event was dispatched but the cache appears to still be empty!"
+            self.log_to_dev_log(msg)
+            return
+
+        self._guild_available.set()
+
+    @abstractmethod
+    async def log_to_dev_log(self, message: str) -> None:
+        """Log the given message to #dev-log."""
+        ...
+
+    async def wait_until_guild_available(self) -> None:
+        """
+        Wait until the guild that matches the ``guild_id`` given at init is available (and the cache is ready).
+
+        The on_ready event is inadequate because it only waits 2 seconds for a GUILD_CREATE
+        gateway event before giving up and thus not populating the cache for unavailable guilds.
+        """
+        await self._guild_available.wait()
+
+    async def setup_hook(self) -> None:
+        """
+        An async init to startup generic services.
+
+        Connects to statsd, and calls
+        :func:`AsyncStatsClient.create_socket <botcore.async_stats.AsyncStatsClient.create_socket>`
+        and :func:`ping_services`.
+        """
+        loop = asyncio.get_running_loop()
+        self._connect_statsd(self.statsd_url, loop)
+        self.stats = AsyncStatsClient(loop, "127.0.0.1")
+        await self.stats.create_socket()
+
+        try:
+            await self.ping_services()
+        except Exception as e:
+            raise StartupError(e)
+
+    @abstractmethod
+    async def ping_services() -> None:
+        """Ping all required services on setup to ensure they are up before starting."""
+        ...
+
+    async def close(self) -> None:
+        """Close the Discord connection, and the aiohttp session, connector, statsd client, and resolver."""
+        # Done before super().close() to allow tasks finish before the HTTP session closes.
+        for ext in list(self.extensions):
+            with suppress(Exception):
+                await self.unload_extension(ext)
+
+        for cog in list(self.cogs):
+            with suppress(Exception):
+                await self.remove_cog(cog)
+
+        # Wait until all tasks that have to be completed before bot is closing is done
+        log.trace("Waiting for tasks before closing.")
+        await asyncio.gather(*self.closing_tasks)
```

Comment:
Maybe there's a better way to deal with this now. Not sure if that would be out of scope for this PR.

Reply:
I blindly copied this from bot and it isn't actually used by anything anymore. I'm tempted to remove it entirely.

Reply:
Only thing I can remember using it was the redis cog. I don't quite remember what the problem even was at this point, but there are comments on the PR that have info.

Reply:
Doing a naive `git log --grep closing_tasks` I found these two:
 - https://github.com/python-discord/bot/commit/429cc865309242f0cf37147f9c3f05036972eb8c - Reddit cog to revoke the access token on unload, which has since been moved to lance (without this feature).
 - https://github.com/python-discord/bot/commit/f4004d814c1babfb5906afb8cd9944ceef90a2a3 - Silence cog, which has been removed since.

Reply:
Removed in https://github.com/python-discord/bot-core/pull/42/commits/aef8b302e89dd84726a950f50650e489d9b9b3b2